docs/tutorial/write-a-machine-charm-for-tinyproxy.md

```diff
@@ -0,0 +1,647 @@
+
+```python
+import logging
+import os
+import re
+import shutil
+import signal
+import subprocess
+
+from charmlibs import apt, pathops
+
+logger = logging.getLogger(__name__)
+
+CONFIG_FILE = "/etc/tinyproxy/tinyproxy.conf"
+PID_FILE = "/var/run/tinyproxy.pid"
+
+
+def check_slug(slug: str) -> None:
+    """Check that the URL slug is valid. Raise ValueError otherwise."""
+    if not re.fullmatch(r"[a-z0-9-]+", slug):
+        raise ValueError(f"Invalid slug: '{slug}'. Slug must match the regex [a-z0-9-]+")
+
+
+def ensure_config(port: int, slug: str) -> bool:
+    """Ensure that tinyproxy is configured. Return True if any changes were made."""
+    # For the config file format, see https://manpages.ubuntu.com/manpages/jammy/en/man5/tinyproxy.conf.5.html
+    config = f"""\
+PidFile "{PID_FILE}"
+Port {port}
+Timeout 600
+ReverseOnly Yes
+ReversePath "/{slug}/" "http://www.example.com/"
+"""
+    return pathops.ensure_contents(CONFIG_FILE, config)
+
+
+def get_version() -> str:
+    """Get the version of tinyproxy that is installed."""
+    result = subprocess.run(["tinyproxy", "-v"], check=True, capture_output=True, text=True)
+    return result.stdout.removeprefix("tinyproxy").strip()
+
+
+def install() -> None:
+    """Use APT to install the tinyproxy executable."""
+    apt.update()
+    apt.add_package("tinyproxy-bin")
+    # If this call fails, the charm will go into error status. The Juju logs will show the error:
+    # charmlibs.apt.PackageError: Failed to install packages: tinyproxy-bin
+
+
+def is_installed() -> bool:
+    """Return whether the tinyproxy executable is available."""
+    return shutil.which("tinyproxy") is not None
+
+
+def is_running() -> bool:
+    """Return whether tinyproxy is running."""
+    return bool(_get_pid())
+
+
+def start() -> None:
+    """Start tinyproxy."""
+    subprocess.run(["tinyproxy"], check=True, capture_output=True, text=True)
+
+
+def stop() -> None:
+    """Stop tinyproxy."""
+    pid = _get_pid()
+    if pid:
+        os.kill(pid, signal.SIGTERM)
+
+
+def reload_config() -> None:
+    """Ask tinyproxy to reload config."""
+    pid = _get_pid()
+    if not pid:
+        raise RuntimeError("tinyproxy is not running")
+    # See https://manpages.ubuntu.com/manpages/jammy/en/man8/tinyproxy.8.html#signals
+    os.kill(pid, signal.SIGUSR1)
+
+
+def _get_pid() -> int | None:
+    """Return the PID of the tinyproxy process, or None if the process can't be found."""
+    if not pathops.LocalPath(PID_FILE).exists():
+        return None
+    pid = int(pathops.LocalPath(PID_FILE).read_text())
+    # Check that we can send signals to the process.
+    try:
+        os.kill(pid, 0)
+    except ProcessLookupError:
+        return None
+    return pid
+```
+
+Notice that the helper module is stateless. In fact, your charm as a whole will be stateless. The main logic of your charm will:
+
+1. Receive an event from Juju. For example, the `start` event, meaning "it's time to start tinyproxy".
```

Comment:
This is sadly incorrect.
"start" means this unit has started.
it may or may not be able to start the workload... because the storage is not attached yet.

See https://discourse.charmhub.io/t/charm-lifecycle/5938